### PR TITLE
Implemented a CostDistribution Class

### PR DESCRIPTION
**Pull request by BenTaylor-TfN**
Describe Changes
---
Adding functionality to support [caf.distribute](https://github.com/Transport-for-the-North/caf.distribute)

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by BenTaylor-TfN** [comment]
@wsp-mbuckley - please can you review? You'll need to understand this code going forward anyway.

I'm just finishing off writing tests, so take them with a pinch of salt.

File: src/caf/toolkit/cost_utils.py [hunk 1]
```diff
@@ -3,25 +3,422 @@
...
+        return self.df[self.trips_col].values
+
+    @property
+    def band_share_vals(self) -> np.ndarray:
+        """Band share values for each of the cost distribution bins."""
+        trip_vals = self.trip_vals
+        return trip_vals / np.sum(trip_vals)
+
+    @staticmethod
+    def from_data(
```
Comment (wsp-mbuckley) [inline src/caf/toolkit/cost_utils.py:151]: This could be a class method then you can access the default values of `min_col`, `max_col`, `avg_col` and `trips_col` instead of redefining them

File: src/caf/toolkit/cost_utils.py [hunk 2]
```diff
@@ -3,25 +3,422 @@
...
+        matrix: np.ndarray,
+        cost_matrix: np.ndarray,
+        min_bounds: Optional[list[float] | np.ndarray] = None,
+        max_bounds: Optional[list[float] | np.ndarray] = None,
+        bin_edges: Optional[list[float] | np.ndarray] = None,
```
Comment (wsp-mbuckley) [inline src/caf/toolkit/cost_utils.py:156]: If you're using `np.histogram` then this can technically be an int for number of bins, with a few changes to this method, is that functionality that is worth enabling?

File: src/caf/toolkit/cost_utils.py [hunk 3]
```diff
@@ -3,25 +3,422 @@
...
+            }
+        )
+
+        return CostDistribution(
+            df=df,
+            min_col=min_col,
+            max_col=max_col,
+            avg_col=avg_col,
+            trips_col=trips_col,
+        )
```
Comment (wsp-mbuckley) [inline src/caf/toolkit/cost_utils.py:224]: If this was a class method

```suggestion
        # Covert data into instance of this class
        df = pd.DataFrame(
            {
                cls.min_col: bin_edges[:-1],
                cls.max_col: bin_edges[1:],
                cls.avg_col: (np.array(bin_edges[:-1]) + np.array(bin_edges[1:])) / 2,
                cls.trips_col: distribution,
            }
        )

        return CostDistribution(df=df)
```

**Review by wsp-mbuckley — CHANGES_REQUESTED** [review]
Looks good, main comments are around making the df attribute private and turning `from_data` into a class method.

File: src/caf/toolkit/cost_utils.py [hunk 4]
```diff
@@ -3,25 +3,416 @@
...
+        Average values for each of the cost distribution bins.
+
+    trip_vals:
+        Trip values for each of the cost distribution bins.
+
+    band_share_vals:
+        Band share values for each of the cost distribution bins.
+    """
+
+    df: pd.DataFrame
```
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:63]: How would that alias work? I tried having a search in the docs, but think I'm looking for the wrong words. If you know off the top of your head, that would help massively!

Once it's private, I'll just create a property for `df` so it can be gotten, but not changed at all.

File: src/caf/toolkit/cost_utils.py (hunk 3 above)
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py]: Solved (alongside a bug or two it created) here: https://github.com/Transport-for-the-North/caf.toolkit/pull/27/commits/473c414343f34e1b61fa21f4b4926d5b7e85014b

File: src/caf/toolkit/cost_utils.py (hunk 2 above)
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:157]: I hadn't thought of it in that way, but it's a valid point.

I think it's something worth considering, but maybe out of scope for this commit. I'll add this as an issue for us to come back to later.

File: src/caf/toolkit/cost_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply src/caf/toolkit/cost_utils.py:63]: Pydantic has a Field class which can be used to define extra information about the attribute including alias, I think this should work `_df: pd.DataFrame = pydantic.Field(alias="df)`, the docs for field are https://docs.pydantic.dev/latest/usage/fields/

[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:63]: This doesn't seem to play nice with the current code. Getting a validation error on constructor call when trying to access the `df`/`_df` item (I tired both just incase).
Is there no way to make `df` and init_only variable?

Using alias also produced PyLint errors everywhere as it doesn't recognise the different name.

[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:63]: https://github.com/pydantic/pydantic/issues/4286 implies that this it to do with the pydantic.dataclass and alias not playing nice together. Only fixed in V2, which breaks the rest of the code

[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:63]: @wsp-mbuckley, are you happy to approve these changes? 

I'll make an issue for this one too, then we can come back to it.

[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:63]: Issue added at #65

File: src/caf/toolkit/cost_utils.py (hunk 2 above)
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:157]: Issue added at #64

**Review by wsp-mbuckley — APPROVED** [review]
